libmcount/dynamic.c

```diff
@@ -376,10 +376,16 @@ static bool match_pattern_list(struct uftrace_mmap *map, char *sym_name,
 	char *libname = basename(map->libname);
 
 	list_for_each_entry(pl, &patterns, list) {
-		if (strncmp(libname, pl->module, strlen(pl->module)))
-			continue;
+		int len = strlen(pl->module);
+		bool matched = false;
+
+		if (soname != NULL)
+			matched = (strncmp(soname, pl->module, len) == 0);
+
+		if (!matched)
+			matched = (strncmp(libname, pl->module, len) == 0);
```

Comment:
How about this?

```c
if (strncmp(libname, pl->module, len) &&
    (!soname || strncmp(soname, pl->module, len)))
        continue;
```

Maybe you can factor out this into a function and call it for dlopen as well.